File: app/src/main/java/com/example/srijith/recyclerviewdemo/type/ReorderItem.java
package com.example.srijith.recyclerviewdemo.type;

public class ReorderItem {

    private int type;

    public ReorderItem(int type) {
        this.type = type;
    }

    public int getType() {
        return type;
    }

    public enum Type {
        PLACEHOLDER(1),
        FAV_HEADER(2),
        OTHER_HEADER(3),
        ITEM_USER(4);

        public int id;

        Type(int id) {
            this.id = id;
        }
    }
}

File: app/src/main/java/com/example/srijith/recyclerviewdemo/type/ReorderItem.kt
package com.example.srijith.recyclerviewdemo.type

open class ReorderItem(val type: Int) {
    enum class Type private constructor(var id: Int) {
        PLACEHOLDER(1),
        FAV_HEADER(2),
        OTHER_HEADER(3),
        ITEM(4)
    }
}

open class Header(val title: String, type: Int) : ReorderItem(type)

class HeaderFavourite : Header("Favourites", Type.FAV_HEADER.id)
class HeaderOthers : Header("Others", Type.OTHER_HEADER.id)
class Placeholder : ReorderItem(Type.PLACEHOLDER.id)

data class ImageItem(val name: String,
                     val imageUrl: String,
                     var buttonType: ButtonType) : ReorderItem(Type.ITEM.id)

enum class ButtonType {
    HIDDEN,
    ADD,
    REMOVE
}